MBS-1870 / MBS-8477 / MBS-10690 / MBS-10691: User tag page improvements

reosarevok opened the pull request:
https://tickets.metabrainz.org/browse/MBS-1870
https://tickets.metabrainz.org/browse/MBS-8477

I know, I know, these should be two (or three) different commits, but I worked on both things at once while converting to React and then it was quite hard to just split them in chunks. Hopefully it's not too hard to review as it is, but I can try to change it if needed (it'll just take a fair amount of time).

yvanzo commented:
I just created ticket [MBS-10341](https://tickets.metabrainz.org/browse/MBS-10341) “Convert user tag entity list page to React” which actually covers the main changes of this PR (if I skimmed through it correctly). It seems impossible to split MBS-10341 from [MBS-1870](https://tickets.metabrainz.org/browse/MBS-1870) as component is paginated. But [MBS-8477](https://tickets.metabrainz.org/browse/MBS-8477) should be easy to commit separately.

reosarevok commented:
@yvanzo: While I agree that ideally I should have split those two, they touch the same files and interact a lot so I really don't feel the (fairly large) amount of time that would go into splitting those would be worth it to be honest. I'll try to avoid overloading commits in the future (well, I'm already trying, but this will help keep it in mind).

yvanzo commented:
Ok but please add a reference to MBS-10341. (That won't make the subject line go over 72 chars :stuck_out_tongue:)

reosarevok commented:
For some test of how this works (better to use the RO Pink DB rather than the sample): /user/reosarevok/tag/to%20clean%20up (per-entity shorter lists + pagination), /user/reosarevok/tag/classical (upvoted vs downvoted).

mwiencek commented on root/user/components/DownvotedSwitch.js:
```diff
@@ -0,0 +1,41 @@
+        {downvotes_link|Show their votes against the tag instead}.`,
+        {downvotes_link: uriWith($c.req.uri, {show_downvoted: 1})},
+      )
+    )}
```
I wonder if we should change the page headings from just `Tag “foo”` to `Votes for tag “foo”` and `Votes against tag “foo”` depending on what's being viewed, instead of explaining it down here. It'd then be nice to make the link text a bit more descriptive here by including the user/tag in the link text, e.g. `"Show {upvotes_link|entities {user} has tagged “{tag}”}.` and `"Show {downvotes_link|entities where {user} has voted against “{tag}”}.`. What do you think? :)

reosarevok replied:
Seems sensible. Done!

reosarevok commented on root/tag/utility/buildTagListSection.js:
```diff
@@ -0,0 +1,85 @@
+import expand2text from '../../static/scripts/common/i18n/expand2text';
+
+
+type Props = {
```
It's not quite the same - this has optional booleans (because they don't exist for TagIndex) but UserTag has mandatory ones. Might not be worth the distinction though, I dunno.

yvanzo commented:
> Hopefully it's not too hard to review as it is

:rofl:

reosarevok commented on root/components/TagList.js:
```diff
@@ -0,0 +1,142 @@
+    */}
+    {buildTagListSection(props, 'area', l('Areas'), N_ln(
+      'See all {num} areas',
+      'See all {num} areas',
```
These aren't new strings, should I still make a change?

yvanzo replied:
Oops, I missed that, never mind.

reosarevok commented:
Added two more commits here, one converting the list of tags by a user to React, another allowing seeing downvoted tags (which wasn't a new issue, FWIW - that was always filtered by is_upvote - but I agree is useful!).

Adding pagination to the list of tags would require very significant changes to how it works, I think - right now it makes one query and then splits it into tags and genres, but that wouldn't work with pagination. We'd need to paginate each list separately, probably requiring moving to two queries. I'd think the number of users who could benefit from pagination on that page is relatively small, so I wouldn't bother for now and would wait until an actual user asks :)

mwiencek commented on root/user/UserTagEntity.js:
```diff
@@ -0,0 +1,68 @@
+        pager={pager}
+      />
+    ) : (
+      <UserHasNotUsedTag
+        entityType={entityType}
+        showDownvoted={showDownvoted}
+        tag={tag}
+        user={user}
+      />
```
Is saying "user has not tagged anything with foo" or "user has not voted against any foo tags" here when they haven't with this particular entity type but *have* with others a bit misleading? At least it appears under e.g. "0 artists found" so it has some context, but dunno if that applies as much for other translations. It might be clearer with only the heading, actually ("0 artists found") and no explanatory text.

reosarevok replied:
I think the explanatory text makes sense, but I made it so that it changes based on entity.

reosarevok commented on root/components/TagList.js:
```diff
@@ -0,0 +1,142 @@
+          <li key="see-all">
+            <em>
+              {user ? (
+                <a
+                  href={'/user/' + encodeURIComponent(user.name) + '/tag/' +
+                        encodeURIComponent(props.tag.name) + '/' + url}
+                >
+                  {expand2text(seeAllMessage(tags.count), {num: tags.count})}
+                </a>
```
Did that on the commit that introduces UserTagLink (also did the same for TagLink below)

reosarevok commented on root/user/components/DownvotedSwitch.js:
```diff
@@ -0,0 +1,87 @@
+  <p>
+    {showDownvoted ? (
+      exp.l(
+        `Show {upvotes_link|entities {user}
```
I made it use the same text as AllDownvotedSwitch: "Showing tags {user} <strong>has voted against</strong>. Show {upvotes_link|entities {user} has tagged with “{tag}”} instead." and whatnot. I think that makes the order seem sensible again :)

reosarevok commented:
Rebased this, fixed a couple small issues I found when testing. Tested by visiting http://localhost:5000/user/reosarevok/tags, http://localhost:5000/user/reosarevok/tag/classical (for up/downvoted switch) and http://localhost:5000/user/reosarevok/tag/to%20clean%20up (for many tagged entities collapsing into a list and see more link, plus per-entity-type lists).

yvanzo commented on root/user/components/UserHasNotUsedTag.js:
```diff
@@ -0,0 +1,106 @@
+  +user: EditorT,
+};
+
+const noDownvotesText = {
+  area: N_l('{user} has not voted against any “{tag}” tags for areas.'),
+  artist: N_l('{user} has not voted against any “{tag}” tags for artists.'),
+  event: N_l('{user} has not voted against any “{tag}” tags for events.'),
+  instrument: N_l(
+    '{user} has not voted against any “{tag}” tags for instruments.',
+  ),
+  label: N_l('{user} has not voted against any “{tag}” tags for labels.'),
+  place: N_l('{user} has not voted against any “{tag}” tags for places.'),
+  recording: N_l(
+    '{user} has not voted against any “{tag}” tags for recordings.',
+  ),
+  release: N_l('{user} has not voted against any “{tag}” tags for releases.'),
+  release_group: N_l(
+    '{user} has not voted against any “{tag}” tags for release groups.',
+  ),
+  series: N_l('{user} has not voted against any “{tag}” tags for series.'),
+  work: N_l('{user} has not voted against any “{tag}” tags for works.'),
+};
+
+const noTagsText = {
+  area: N_l('{user} has not tagged any areas with “{tag}”.'),
+  artist: N_l('{user} has not tagged any artists with “{tag}”.'),
+  event: N_l('{user} has not tagged any events with “{tag}”.'),
+  instrument: N_l('{user} has not tagged any instruments with “{tag}”.'),
+  label: N_l('{user} has not tagged any labels with “{tag}”.'),
+  place: N_l('{user} has not tagged any places with “{tag}”.'),
+  recording: N_l('{user} has not tagged any recordings with “{tag}”.'),
+  release: N_l('{user} has not tagged any releases with “{tag}”.'),
+  release_group: N_l(
+    '{user} has not tagged any release groups with “{tag}”.',
+  ),
+  series: N_l('{user} has not tagged any series with “{tag}”.'),
+  work: N_l('{user} has not tagged any works with “{tag}”.'),
+};
```
Please provide a URL example using these strings, if any.

reosarevok replied:
http://localhost:5000/user/reosarevok/tag/classical/release /

yvanzo replied:
Thanks, how do you reach these pages?

reosarevok replied:
I edited the comment in the meantime to specify that, sorry! :)

yvanzo replied:
Found it: You click on “See all {num} works” under the short list of 10 first works available from https://musicbrainz.org/user/jesus2099/tag/instrumental

So there seems to be two issues:
* Navigation: There should be a way to go back to `/tag/instrumental` from `/tag/instrumental/work`.
* Description: Why are these strings entity type-specific and neither the header `Vote for/against “{tag}”` nor its following paragraph? If one of these strings should be entity type-specific that should probably be the header, other strings can probably be entity type-neutral, like what we have for subscriptions already.

reosarevok replied:
Both should be, because just saying "jesus2099 has not voted against any “instrumental” tags" would be a lie :) But that's fair, other lines here should also be specific, at least the "Showing entities where" bit.

yvanzo reviewed (changes requested):
Please write user-friendly descriptions of linked tickets so they can find affected pages for them.

Additionally tested with:
* http://localhost:5000/user/reosarevok/tags?show_downvoted=1
* http://localhost:5000/user/reosarevok/tag/classical?show_downvoted=1
* http://localhost:5000/user/reosarevok/tag/to%20clean%20up?show_downvoted=1
* http://localhost:5000/user/reosarevok/tag/calypso (for tag not used by an editor)
* http://localhost:5000/user/reosarevok/tag/calypso?show_downvoted=1 (for tag not voted against by an editor)

I did not find any use of entity type-specific messages yet. Are they really used anywhere?

reosarevok commented:
Updated the tickets a bit with some examples - do check if you think they look good! :)

mwiencek commented:
I reviewed the recent additions here and decided to play around with some UI changes, because I wasn't too happy with the number of strings being added. The UI also isn't very easy to use despite all these strings being added. I'm no designer, but can think of a few reasons...

  * I have to read several complete sentences to decide what I want to do or where I want to navigate. (As opposed to having a breadcrumb or simple dropdown.)
  * The strings repeat each other, or are too specific without adding any new information. I think the "no results" page illustrates that best: 
    <img width="831" alt="Example of repetitive strings on the no results page" src="https://user-images.githubusercontent.com/1056556/110515411-ff9c0c80-80cd-11eb-9261-f083eed8a2e6.png">
    The header and first sentence here basically repeat each other. So do the last two sentences. I'm not sure we need a string like "x has not tagged any areas with y" when "0 areas found" suffices. It's obvious from the context of the page what "0 areas found" means: that we found 0 areas the user tagged with y.
  * The terminology feels inconsistent. Across all of these pages, I see: "tagged," "uses of tag," "has used," "voted for," "voted against."

Since I don't want you to have to rework this PR again, I pushed a "WIP" commit here that I hope we could work off of. It adds a "vote filter" select, breadcrumbs, and consolidates a lot of the strings. It's not perfect, but closer to what I think things should be. Can you have a look and see if the changes make sense?

<img width="385" alt="Example of the new no results page" src="https://user-images.githubusercontent.com/1056556/110519270-b8fce100-80d2-11eb-973a-0e13b3f74a24.png">
<img width="333" alt="Example of the user tagged entities overview page" src="https://user-images.githubusercontent.com/1056556/110519282-bc906800-80d2-11eb-8f38-bb4d194cdccb.png">

reosarevok commented:
Having filter *and* an "Upvoted tags" link such as below is very confusing IMO.
![Screenshot from 2021-03-10 12-54-01](https://user-images.githubusercontent.com/1069224/110618777-bddb9680-819f-11eb-8cec-ca6bc6fe0b6f.png)

I understand it's meant as a breadcrumb for "All upvoted tags", but maybe that could be shown either on top or as part of the header? The other page we use breadcrumbs on that I can think of is admin/attributes (https://musicbrainz.org/admin/attributes/WorkAttributeType for example). There we use the header, but I guess the main point of a separate header here is that it allows us to link to the tag. So maybe on top is better.

I'd also consider using "Show:" instead of "Vote filter:", and ideally having that under the heading.

For an example: 
![Screenshot from 2021-03-10 13-13-45](https://user-images.githubusercontent.com/1069224/110621095-815d6a00-81a2-11eb-8176-7619e61fd879.png)

yvanzo reviewed (approved):
Tested with the above URLs again. The breadcrumb improves navigation a lot. Plus it skips adding too many strings! :rocket:
